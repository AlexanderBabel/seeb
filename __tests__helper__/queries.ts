export const todosQuery = `{
  todos {
    id
    name
    completed
    dueDate
  }
}`;

export const todoQuery = `{
  todo(id: "we121ew") {
    id
    name
    completed
    dueDate
  }
}`;

export const createTodoQuery = `
mutation createTodo {
  createTodo(name: "Test #2") {
    id
    name
    completed
    dueDate
  }
}`;

export const createTodo2Query = `
mutation createTodo {
  createTodo(name: "Test #2", dueDate: "2020-01-01T00:00:00.000Z") {
    id
    name
    completed
    dueDate
  }
}`;
