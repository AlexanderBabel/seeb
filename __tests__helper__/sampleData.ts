import { Todo } from '../src/graphql/typeDefs/Todo';

export const TodoData: Todo = {
  id: 'we121ew',
  name: 'Test Todo',
  completed: false,
};

export const TodoData2: Todo = {
  id: 'hr848rh',
  name: 'Todo #2',
  completed: true,
  dueDate: new Date('2020-01-01T00:00:00.000Z'),
};

export const TodosData: Todo[] = [TodoData, TodoData2];
