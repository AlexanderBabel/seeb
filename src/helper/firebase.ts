import admin from 'firebase-admin';
import { FirebaseTodo } from '../types/todo';

admin.initializeApp();
const { fromDate } = admin.firestore.Timestamp;
export const collection = admin.firestore().collection('todos');
export const firestore = admin.firestore();

export async function addTodo(name: string, dueDate?: Date): Promise<FirebaseTodo> {
  const data = {
    name,
    completed: false,
    dueDate: dueDate ? fromDate(dueDate) : null,
  };

  const res = await collection.add(data);
  return {
    ...data,
    id: res.id,
    dueDate,
  };
}

export async function deleteTodo(id: string): Promise<boolean> {
  const doc = firestore.doc(`todos/${id}`);

  return await doc.delete().then(() => true);
}

export async function getTodos(): Promise<FirebaseTodo[]> {
  return Promise.all(
    (await collection.get()).docs.map(async (d) => {
      const data = await d.data();
      return {
        id: d.id,
        ...data,
        dueDate: data.dueDate ? (data.dueDate as admin.firestore.Timestamp).toDate() : null,
      } as FirebaseTodo;
    })
  );
}
