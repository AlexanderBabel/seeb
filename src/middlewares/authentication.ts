import { Request, NextFunction, Response } from 'express';
import { verify } from 'jsonwebtoken';

const allowedRoutes = ['/api-docs*', '/token'];

export type Context = {
  user: string;
};

declare global {
  namespace Express {
    interface Request {
      context: Context;
    }
  }
}

export default (req: Request, res: Response, next: NextFunction): void => {
  const isGraphQL = req.path.match('/graphql');
  if (allowedRoutes.find((r) => req.path.match(r))) {
    return next();
  }

  const { authorization } = req.headers;
  const { JWT_SECRET } = process.env;
  if (!JWT_SECRET) {
    res.status(500).send('Missing env var.');
    return;
  }

  if (!authorization) {
    if (isGraphQL) {
      return next();
    }

    res.status(401).send('Not Authenticated');
    return;
  }

  try {
    const decoded = verify(authorization.replace('Bearer ', ''), JWT_SECRET) as { user: string };
    if (typeof decoded === 'object' && decoded.user) {
      req.context = { user: decoded.user };
    }
  } catch (err) {
    if (!isGraphQL) {
      res.status(401).send('Not Authenticated');
      return;
    }
  }

  return next();
};
