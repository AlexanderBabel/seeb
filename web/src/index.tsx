import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";

import AccountProvider from "./contexts/AccountContext";

import RegisteredRoute from "./components/RegisteredRoute";
import UnregisteredRoute from "./components/UnregisteredRoute";
import Register from "./screens/Register";
import Home from "./screens/Home";
import Error404 from "./screens/Error404";

import * as serviceWorker from "./serviceWorker";

import "./styles/tailwind.out.css";
import { Apollo } from "./contexts/Apollo";

ReactDOM.render(
  <React.StrictMode>
    <AccountProvider>
      <Apollo>
        <Router>
          <Switch>
            <UnregisteredRoute path="/register" exact component={Register} />
            <RegisteredRoute path="/" exact component={Home} />
            <Route component={Error404} />
          </Switch>
        </Router>
      </Apollo>
    </AccountProvider>
  </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();
