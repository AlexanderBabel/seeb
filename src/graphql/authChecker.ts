import { AuthChecker } from 'type-graphql';
import { Context } from '../middlewares/authentication';

export const authChecker: AuthChecker<Context> = ({ context }) => {
  return !!context.user;
};
