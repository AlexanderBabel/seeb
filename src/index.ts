import bodyParser from 'body-parser';
import cors from 'cors';
import express from 'express';
import helmet from 'helmet';
import 'dotenv/config';

import authentication from './middlewares/authentication';
import createTodo from './routes/createTodo';
import deleteTodo from './routes/deleteTodo';
import getTodos from './routes/getTodos';
import updateTodo from './routes/updateTodo';
import createFirebaseTodo from './routes/firebase/createFirebaseTodo';
import getFirebaseTodos from './routes/firebase/getFirebaseTodos';

const app = express();

// middlewares
app.use(cors());
app.use(helmet());
app.use(bodyParser.json());
app.use(authentication);

// routes
app.get('/todo', getTodos);
app.post('/todo', createTodo);
app.patch('/todo', updateTodo);
app.delete('/todo', deleteTodo);

app.get('/firebase/todo', getFirebaseTodos);
app.post('/firebase/todo', createFirebaseTodo);

app.listen(process.env.PORT ?? 4000, () => {
  console.log(`Started server on port ${process.env.PORT ?? 4000}`);
});
