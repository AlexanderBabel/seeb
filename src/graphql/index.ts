import 'reflect-metadata';
import { Express } from 'express';
import { ApolloServer } from 'apollo-server-express';
import { TodoResolver } from './resolvers/TodoResolver';
import { buildSchema, Ctx } from 'type-graphql';
import { authChecker } from './authChecker';
import { GraphQLSchema } from 'graphql';

export function getSchema(): Promise<GraphQLSchema> {
  return buildSchema({
    resolvers: [TodoResolver],
    authChecker,
  });
}

export default async function (app: Express): Promise<void> {
  const server = new ApolloServer({
    schema: await getSchema(),
    context: (ctx) => ctx.req.context,
  });

  server.applyMiddleware({ app });
}
