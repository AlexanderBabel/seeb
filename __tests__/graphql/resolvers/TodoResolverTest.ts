import { GraphQLSchema, graphql } from 'graphql';
import { getSchema } from '../../../src/graphql';
import { Context } from '../../../src/middlewares/authentication';
import * as firebase from '../../../src/helper/firebase';
import { mocked } from 'ts-jest/utils';
import { TodosData, TodoData, TodoData2 } from '../../../__tests__helper__/sampleData';
import {
  todosQuery,
  todoQuery,
  createTodoQuery,
  createTodo2Query,
} from '../../../__tests__helper__/queries';

jest.mock('../../../src/helper/firebase');

const firebaseMock = mocked(firebase);
firebaseMock.addTodo.mockResolvedValue(TodoData);
firebaseMock.getTodo.mockResolvedValue(TodoData);

let schema: GraphQLSchema;
const ctx: Context = { user: 'JEST_TEST' };

beforeAll(async (done) => {
  schema = await getSchema();
  done();
});

describe('Test of TodoResolver', () => {
  test('Test of "todo" query', async (done) => {
    firebaseMock.getTodo.mockResolvedValue(TodoData);

    const res = await graphql(schema, todoQuery, null, ctx);
    expect(res).toMatchSnapshot();

    expect(firebaseMock.getTodo).toHaveBeenCalledTimes(1);
    expect(firebaseMock.getTodo.mock.calls[0]).toMatchSnapshot();

    firebaseMock.getTodo.mockReset();
    done();
  });

  test('Test of "todo" query with no result', async (done) => {
    firebaseMock.getTodo.mockResolvedValue(undefined);

    const res = await graphql(schema, todoQuery, null, ctx);
    expect(res).toMatchSnapshot();

    expect(firebaseMock.getTodo).toHaveBeenCalledTimes(1);
    expect(firebaseMock.getTodo.mock.calls[0]).toMatchSnapshot();

    firebaseMock.getTodo.mockReset();
    done();
  });

  test('Test of "todos" query', async (done) => {
    firebaseMock.getTodos.mockResolvedValue(TodosData);

    const res = await graphql(schema, todosQuery, null, ctx);
    expect(res).toMatchSnapshot();

    expect(firebaseMock.getTodos).toHaveBeenCalledTimes(1);
    expect(firebaseMock.getTodos.mock.calls[0]).toMatchSnapshot();

    firebaseMock.getTodos.mockReset();
    done();
  });

  test('Test of "createTodo" mutation without dueDate', async (done) => {
    firebaseMock.addTodo.mockResolvedValue(TodoData2);

    const res = await graphql(schema, createTodoQuery, null, ctx);
    expect(res).toMatchSnapshot();

    expect(firebaseMock.addTodo).toHaveBeenCalledTimes(1);
    expect(firebaseMock.addTodo.mock.calls[0]).toMatchSnapshot();

    firebaseMock.addTodo.mockReset();
    done();
  });

  test('Test of "createTodo" mutation with dueDate', async (done) => {
    firebaseMock.addTodo.mockResolvedValue(TodoData2);

    const res = await graphql(schema, createTodo2Query, null, ctx);
    expect(res).toMatchSnapshot();

    expect(firebaseMock.addTodo).toHaveBeenCalledTimes(1);
    expect(firebaseMock.addTodo.mock.calls[0]).toMatchSnapshot();

    firebaseMock.addTodo.mockReset();
    done();
  });
});
