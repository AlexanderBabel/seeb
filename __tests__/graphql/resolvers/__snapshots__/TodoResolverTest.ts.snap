// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of TodoResolver Test of "createTodo" mutation with dueDate 1`] = `
Object {
  "data": Object {
    "createTodo": Object {
      "completed": true,
      "dueDate": "2020-01-01T00:00:00.000Z",
      "id": "hr848rh",
      "name": "Todo #2",
    },
  },
}
`;

exports[`Test of TodoResolver Test of "createTodo" mutation with dueDate 2`] = `
Array [
  "JEST_TEST",
  "Test #2",
  2020-01-01T00:00:00.000Z,
]
`;

exports[`Test of TodoResolver Test of "createTodo" mutation without dueDate 1`] = `
Object {
  "data": Object {
    "createTodo": Object {
      "completed": true,
      "dueDate": "2020-01-01T00:00:00.000Z",
      "id": "hr848rh",
      "name": "Todo #2",
    },
  },
}
`;

exports[`Test of TodoResolver Test of "createTodo" mutation without dueDate 2`] = `
Array [
  "JEST_TEST",
  "Test #2",
  undefined,
]
`;

exports[`Test of TodoResolver Test of "todo" query 1`] = `
Object {
  "data": Object {
    "todo": Object {
      "completed": false,
      "dueDate": null,
      "id": "we121ew",
      "name": "Test Todo",
    },
  },
}
`;

exports[`Test of TodoResolver Test of "todo" query 2`] = `
Array [
  "we121ew",
  "JEST_TEST",
]
`;

exports[`Test of TodoResolver Test of "todo" query with no result 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field Query.todo.],
  ],
}
`;

exports[`Test of TodoResolver Test of "todo" query with no result 2`] = `
Array [
  "we121ew",
  "JEST_TEST",
]
`;

exports[`Test of TodoResolver Test of "todos" query 1`] = `
Object {
  "data": Object {
    "todos": Array [
      Object {
        "completed": false,
        "dueDate": null,
        "id": "we121ew",
        "name": "Test Todo",
      },
      Object {
        "completed": true,
        "dueDate": "2020-01-01T00:00:00.000Z",
        "id": "hr848rh",
        "name": "Todo #2",
      },
    ],
  },
}
`;

exports[`Test of TodoResolver Test of "todos" query 2`] = `
Array [
  "JEST_TEST",
]
`;
